src/resources/ble_esp32.cc

```diff
@@ -697,6 +918,112 @@ PRIMITIVE(request_attribute) {
   return process->program()->null_object();
 }
 
+/*
+ *
+ * Primitives for BLE server
+ *
+ */
+PRIMITIVE(server_config_init) {
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (!proxy) ALLOCATION_FAILED;
+
+  // Mark usage. When the group is unregistered, the usage is automatically
+  // decremented, but if group allocation fails, we manually call unuse().
+  BLEEventSource* ble = BLEEventSource::instance();
+  if (!ble->use()) {
+    MALLOC_FAILED;
+  }
+
+  BLEServerConfigGroup* group = _new BLEServerConfigGroup(process, ble);
+  if (!group) MALLOC_FAILED;
+
+  proxy->set_external_address(group);
+  return proxy;
+}
+
+PRIMITIVE(add_server_service) {
+  ARGS(BLEServerConfigGroup, group, Blob, uuid_blob);
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (!proxy) ALLOCATION_FAILED;
+
+  ble_uuid_any_t uuid = uuid_from_blob(uuid_blob);
+
+  BLEServerServiceResource* service = group->add_service(uuid);
+  if (!service) MALLOC_FAILED;
+
+  proxy->set_external_address(service);
+  return proxy;
+}
+
+PRIMITIVE(add_server_characteristic) {
+  ARGS(BLEServerServiceResource, service, Blob, uuid_blob, int, type, Object, value);
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (!proxy) ALLOCATION_FAILED;
+
+  ble_uuid_any_t uuid = uuid_from_blob(uuid_blob);
+
+  struct os_mbuf* om =null;
+  if (!object_to_mbuf(process, value, &om)) WRONG_TYPE;
+
+  BLEServerCharacteristicResource* characteristic = service->add_characteristic(uuid, type, om);
+  if (!characteristic) MALLOC_FAILED;
+
+  proxy->set_external_address(characteristic);
+  return proxy;
+}
+
+PRIMITIVE(set_characteristics_value) {
+  ARGS(BLEServerCharacteristicResource, resource, Object, value);
+
+  struct os_mbuf* om;
+  if (!object_to_mbuf(process, value, &om)) WRONG_TYPE;
+
+  resource->set_mbuf_to_send(om);
+
+  return process->program()->null_object();
+}
+
+PRIMITIVE(notify_characteristics_value) {
+  ARGS(BLEServerCharacteristicResource, resource, Object, value);
+
+  if (resource->is_notify_enabled() || resource->is_indicate_enabled()) {
+    struct os_mbuf* om;
+    if (!object_to_mbuf(process, value, &om)) WRONG_TYPE;
+
+    int err = 0;
+    if (resource->is_notify_enabled()) {
+      err = ble_gattc_notify_custom(resource->conn_handle(), resource->nimble_value_handle(), om);
+    }
+
+    if (err == 0 && resource->is_indicate_enabled()) {
```

Comment:
Use ESP_OK?

Reply:
It is a bit more complicated. The nimble api does not use the ESP error codes, but its own set. So even though ESO_OK is 0, it would not really be correct to use that constant here. The nimble stack does not have a corresponding BLE_OK, so i kept it as 0.